Store journal abbreviation lists in specified directory

adhamahmad opened the pull request:
Closes https://github.com/JabRef/jabref/issues/10557

### UI changes:
![image](https://github.com/user-attachments/assets/72276c58-07f0-4f3a-b4b8-9f476bc24047)


### Mandatory checks

<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

- [x] I own the copyright of the code submitted and I licence it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if change is visible to the user)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

koppor commented:
Thank you for working on this.

> ### TODO (Not implemented in this PR):

What does this mean?

Will there be follow-up PRs? Do you intent to work on this in the context of commits to this PR?

adhamahmad commented:
@koppor 

 I intend to include them as additional commits within this PR.

koppor commented:
>     * Implementation of .csv to .mv store conversion

Code hint: This is already implemented at org.jabref.cli.JournalListMvGenerator. One just needs to move that class to `org.jabref.logic.journals`  and adapt it.

koppor commented:
>     * Implement automatic detection of journal lists in the specified directory:

This could be easy.

1. Have a HashMap (stored inMVStore) from filename to timestamp
2. Files.walktree -> for each file: look up. if not found: import and put timestamp of file. If found: compare timestamp. If not equal: clear current hashmap for that. import and update time stamp

koppor commented:
@adhamahmad I think, the TODO list in the PR description should be updated if I read your last commit comment right?

adhamahmad commented:
> @adhamahmad I think, the TODO list in the PR description should be updated if I read your last commit comment right?

You're right—the .csv to .mv conversion is now implemented. I updated the TODO list, the only missing part is the unit test.

koppor commented:
This PR might have conflicts with https://github.com/JabRef/jabref/pull/12606 - They should be combined at some point.

Depening on which PR gets in first, the other one needs to be adapted.

koppor commented on src/main/java/org/jabref/gui/preferences/journals/JournalAbbreviationsTabViewModel.java:
```diff
@@ -215,6 +248,38 @@ public void removeCurrentFile() {
+        for (String filePath : externalLists) {
+            if (journalFiles.stream().noneMatch(file -> file.getAbsolutePath().map(Path::toString).orElse("").equals(filePath))) {
+                Path path = Path.of(filePath);
+                if (filePath.endsWith(".mv")) {
+                    openMvFile(path);
+                } else {
+                    openFile(path);
+                }
```
I think, the logic is wrong.

Think of `.mv` as cache.

- If `.csv` is older than `.mv`: Use .mv
- Else: Create .mv from csv

adhamahmad replied:
> I think, the logic is wrong.
> 
> Think of `.mv` as cache.
> 
> * If `.csv` is older than `.mv`: Use .mv
> * Else: Create .mv from csv

I reviewed the logic and found that the else branch (handling CSV files) is not needed.

By the time `updateJournalFiles()` is executed, the conversion process in `JournalAbbreviationMvGenerator.convertAllCsvToMv()` has already performed a timestamp check on all CSV files. 

This ensures that any CSV file in the new directory has been converted to an MV file.

koppor commented:
I think there was something wrong while merging upstream/main. I see old versions of files coming in.

adhamahmad commented:
> I think there was something wrong while merging upstream/main. I see old versions of files coming in.

You're right @koppor , there was an issue while merging upstream/main, which caused old versions of some files to come in.

I realized this happened because I staged unintended changes, likely due to using git add . from the CLI. 

To fix it, I reset my branch to upstream/main, reapplied only my intended changes, and pushed the correct updates.

InAnYan commented on src/main/java/org/jabref/gui/preferences/journals/AbbreviationsFileViewModel.java:
```diff
@@ -107,4 +133,8 @@ public boolean equals(Object obj) {
         }
     }
+
+    public boolean isMvFile() {
```
Can you move this method to `FileUtil`? Similar to this: <https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/logic/util/io/FileUtil.java#L494>

InAnYan commented on src/main/java/org/jabref/gui/preferences/journals/JournalAbbreviationsTabViewModel.java:
```diff
@@ -215,6 +248,36 @@ public void removeCurrentFile() {
+        for (String filePath : externalLists) {
+            if (journalFiles.stream().noneMatch(file -> file.getAbsolutePath().map(Path::toString).orElse("").equals(filePath))) {
+                Path path = Path.of(filePath);
+                if (filePath.endsWith(".mv")) {
```
Make a new method in `FileUtil` like this: <https://github.com/JabRef/jabref/blob/main/src/main/java/org/jabref/logic/util/io/FileUtil.java#L494>.

InAnYan commented on src/main/java/org/jabref/logic/journals/JournalAbbreviationMvGenerator.java:
```diff
@@ -0,0 +1,149 @@
+     */
+    public static void convertAllCsvToMv(Path abbreviationsDirectory) {
+        // A set of filenames to ignore
+        Set<String> ignoredNames = Set.of(
```
Introduce a constant for `ignoredNames`

InAnYan commented on src/main/java/org/jabref/logic/journals/JournalAbbreviationMvGenerator.java:
```diff
@@ -0,0 +1,149 @@
+        );
+
+        // Open or create a persistent MVStore file for storing CSV file timestamps.
+        Path timestampFile = abbreviationsDirectory.resolve("timestamps.mv");
```
Introduce constant for `"timestamps.mv"`

InAnYan commented on src/main/java/org/jabref/logic/journals/JournalAbbreviationMvGenerator.java:
```diff
@@ -0,0 +1,149 @@
+                .fileName(timestampFile.toString())
+                .compressHigh()
+                .open()) {
+            MVMap<String, Long> timestampMap = store.openMap("fileTimestamps");
```
Introduce constant for `"fileTimestampts"`

InAnYan commented on src/main/java/org/jabref/logic/journals/JournalAbbreviationMvGenerator.java:
```diff
@@ -0,0 +1,149 @@
+            LOGGER.error("Error while processing abbreviation files in directory: {}", abbreviationsDirectory, e);
+        }
+    }
+    /**
```
JavaDoc misaligned

InAnYan commented on src/main/java/org/jabref/logic/journals/JournalAbbreviationMvGenerator.java:
```diff
@@ -0,0 +1,149 @@
+                .fileName(path.toString())
+                .readOnly()
+                .open()) {
+            MVMap<String, Abbreviation> abbreviationMap = store.openMap("FullToAbbreviation");
```
`"FullAbbreviation"` should also be a constant

InAnYan commented on src/main/java/org/jabref/logic/journals/JournalAbbreviationMvGenerator.java:
```diff
@@ -0,0 +1,149 @@
+                .fileName(mvFile.toString())
+                .compressHigh()
+                .open()) {
+            MVMap<String, Abbreviation> fullToAbbreviation = store.openMap("FullToAbbreviation");
```
`"FullAbbreviation"` should also be a constant

InAnYan commented on src/main/java/org/jabref/logic/preferences/JabRefCliPreferences.java:
```diff
@@ -673,6 +676,9 @@ protected JabRefCliPreferences() {
         // endregion
+
+        // Journal abbreviations directory
+        defaults.put(JOURNAL_ABBREVIATION_DIRECTORY, Directories.getJournalAbbreviationsDirectory().toString());
```
`default.put` calls are grouped into regions.

There is no region for `JOURNAL_ABBREVIATION_DIRECTORY`?

InAnYan commented on src/main/java/org/jabref/logic/preferences/JabRefCliPreferences.java:
```diff
@@ -577,7 +580,7 @@ protected JabRefCliPreferences() {
         defaults.put(KEEP_DOWNLOAD_URL, Boolean.TRUE);
         defaults.put(DEFAULT_CITATION_KEY_PATTERN, "[auth][year]");
-        defaults.put(UNWANTED_CITATION_KEY_CHARACTERS, "-`ʹ:!;?^$");
+        defaults.put(UNWANTED_CITATION_KEY_CHARACTERS, "-`ʹ:!;?^");
```
Should this change be made? (Maybe I miss some context)

InAnYan commented on src/main/java/org/jabref/logic/preferences/JabRefCliPreferences.java:
```diff
@@ -1012,13 +1018,19 @@ public JournalAbbreviationPreferences getJournalAbbreviationPreferences() {
                 getStringList(EXTERNAL_JOURNAL_LISTS),
-                getBoolean(USE_AMS_FJOURNAL));
+                getBoolean(USE_AMS_FJOURNAL),
+               new SimpleStringProperty(get(JOURNAL_ABBREVIATION_DIRECTORY)));
```
No, this is not how we write preferences in JabRef.

You shouldn't make `Property` classes there

InAnYan commented on src/main/java/org/jabref/logic/journals/JournalAbbreviationPreferences.java:
```diff
@@ -1,23 +1,96 @@
         this.useFJournalField = new SimpleBooleanProperty(useFJournalField);
     }
 
+    public void updateJournalsDir(String directory) {
```
This is `Preferences` class.

Typically, these classes do not contain any custom logic. The only thing it does is to make `Property` objects and provide access to them. No more, no less.

IDK what should be done here, but I just wanted to note that logic isn't really written in `Preferences`-like classes

InAnYan commented on src/main/java/org/jabref/logic/journals/JournalAbbreviationPreferences.java:
```diff
@@ -1,23 +1,96 @@
     public JournalAbbreviationPreferences(List<String> externalJournalLists,
-                                          boolean useFJournalField) {
+                                          boolean useFJournalField,
+                                          StringProperty journalsDir) {
```
We use primitives in `Preferences`-like classes constructors

InAnYan commented on src/main/java/org/jabref/logic/journals/JournalAbbreviationPreferences.java:
```diff
@@ -1,23 +1,96 @@
+                                          boolean useFJournalField,
+                                          StringProperty journalsDir) {
+
+        if (journalsDir == null || journalsDir.get() == null) {
```
Beware of `null`s.

If something can be not present, use Java `Optional`s

jabref-machine commented:
Note that your PR will not be reviewed/accepted until you have gone through the mandatory checks in the description and marked each of them them exactly in the format of `[x]` (done), `[ ]` (not done yet) or `[/]` (not applicable).


<!-- ghprcomment
Tests
PR checklist OK
-->

jabref-machine commented:
JUnit tests are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Tests
Unit tests
-->

jabref-machine commented:
JUnit tests are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Tests
Unit tests
-->

jabref-machine commented:
JUnit tests are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Tests
Unit tests
-->

jabref-machine commented:
JUnit tests are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Tests
Unit tests
-->

jabref-machine commented:
JUnit tests are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Tests
Unit tests
-->

jabref-machine commented:
JUnit tests are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Tests
Unit tests
-->

jabref-machine commented:
JUnit tests are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Tests
Unit tests
-->

jabref-machine commented:
JUnit tests are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Tests
Unit tests
-->

zikunz commented:
This is really cool! I guess if Multi-List Management for JabRef is chosen as a GSoC project this summer, the selected contributor might want to take a look at this PR to continue building on these new changes (i.e., ‘Change Directory’ feature).

adhamahmad commented:
> This is really cool! I guess if Multi-List Management for JabRef is chosen as a GSoC project this summer, the selected contributor might want to take a look at this PR to continue building on these new changes (i.e., ‘Change Directory’ feature).

This feature is indeed part of the broader Multi-List Management concept I’ve been exploring. I’m looking forward to potentially building upon this as part of the GSoC project if I’m selected!

jabref-machine commented:
Your code currently does not meet [JabRef's code guidelines](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). We use [Checkstyle](https://checkstyle.sourceforge.io/) to identify issues. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Tests / Checkstyle (pull_request)" and click on it.

In case of issues with the import order, double check that you [activated Auto Import](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#enable-proper-import-cleanup). You can trigger fixing imports by pressing <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> to trigger [Optimize Imports](https://www.jetbrains.com/guide/tips/optimize-imports/).

Please carefully follow [the setup guide for the codestyle](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). Afterwards, please [run checkstyle locally](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#run-checkstyle) and fix the issues, commit, and push.


<!-- ghprcomment
Tests
Checkstyle
-->
